src/logic/hydra.lean

```diff
@@ -0,0 +1,181 @@
+-/
+
+namespace relation
+
+variables {α β : Type*}
+
+section two_rels
+variables (rα : α → α → Prop) (rβ : β → β → Prop) (f : α → β)
+
+/-- A function `f : α → β` is a fibration between the relation `rα` and `rβ` if for all
+  `a : α` and `b : β`, whenever `b : β` and `f a` are related by `rβ`, `b` is the image
+  of some `a' : α` under `f`, and `a'` and `a` are related by `rα`. -/
+def fibration := ∀ ⦃a b⦄, rβ b (f a) → ∃ a', rα a' a ∧ f a' = b
+
+variables {rα rβ}
+/-- If `f : α → β` is a fibration between relations `rα` and `rβ`, and `a : α` is
+  accessible under `rα`, then `f a` is accessible under `rβ`. -/
+lemma _root_.acc.of_fibration (fib : fibration rα rβ f) {a} (ha : acc rα a) : acc rβ (f a) :=
+begin
+  induction ha with a ha ih,
+  refine acc.intro (f a) (λ b hr, _),
+  obtain ⟨a', hr', rfl⟩ := fib hr,
+  exact ih a' hr',
+end
+
+lemma _root_.acc.of_downward_closed (dc : ∀ {a b}, rβ b (f a) → b ∈ set.range f)
+  (a : α) (ha : acc (inv_image rβ f) a) : acc rβ (f a) :=
+ha.of_fibration f (λ a b h, let ⟨a', he⟩ := dc h in ⟨a', he.substr h, he⟩)
+
+variables (rα rβ)
+/-- The "addition of games" relation in combinatorial game theory, on the product type: if
+  `rα a' a` means that `a ⟶ a'` is a valid move in game `α`, and `rβ b' b` means that `b ⟶ b'`
+  is a valid move in game `β`, then `game_add rα rβ` specifies the valid moves in the juxtaposition
+  of `α` and `β`: the player is free to choose one of the games and make a move in it,
+  while leaving the other game unchanged. -/
+inductive game_add : α × β → α × β → Prop
+| fst {a' a b} : rα a' a → game_add (a',b) (a,b)
+| snd {a b' b} : rβ b' b → game_add (a,b') (a,b)
+
+/-- `game_add` is a `subrelation` of `prod.lex`. -/
+lemma game_add_le_lex : game_add rα rβ ≤ prod.lex rα rβ :=
+λ _ _ h, h.rec (λ _ _ b, prod.lex.left b b) (λ a _ _, prod.lex.right a)
+
+/-- `prod.rprod` is a subrelation of the transitive closure of `game_add`. -/
+lemma rprod_le_trans_gen_game_add : prod.rprod rα rβ ≤ trans_gen (game_add rα rβ) :=
+λ _ _ h, h.rec $ begin
+  intros _ _ _ _ hα hβ,
+  exact trans_gen.tail (trans_gen.single $ game_add.fst hα) (game_add.snd hβ),
+end
+
+variables {rα rβ}
+/-- If `a` is accessible under `rα` and `b` is accessible under `rβ`, then `(a, b)` is
+  accessible under `(a, b)`. Notice that `prod.lex_accessible` requires the stronger
+  condition `∀ b, acc rb b`. -/
+lemma _root_.acc.game_add {a b} (ha : acc rα a) (hb : acc rβ b) : acc (game_add rα rβ) (a, b) :=
+begin
+  induction ha with a ha iha generalizing b,
+  induction hb with b hb ihb,
+  refine acc.intro _ (λ h, _),
+  rintro (⟨_,_,_,ra⟩|⟨_,_,_,rb⟩),
+  exacts [iha _ ra (acc.intro b hb), ihb _ rb],
+end
+
+/-- The addition of two well-founded games is well-founded. -/
+lemma _root_.well_founded.game_add (hα : well_founded rα) (hβ : well_founded rβ) :
+  well_founded (game_add rα rβ) := ⟨λ ⟨a,b⟩, (hα.apply a).game_add (hβ.apply b)⟩
+
+end two_rels
+
+section hydra
+open game_add multiset
+
+variable (r : α → α → Prop)
```

Comment:
This shouldn't be in a `variable` block. Some of the theorems and defs below require it explicit, while others can have it implicit.

Reply:
I made it implicit for `cut_expand_iff` and all declarations since `acc_of_singleton`. You just write `variable {r}` when you want it implicit and doesn't need to write the type ever again.